### ISSUE TITLE
some upload and download problem

### ISSUE THREAD

**Issue by tlightsky** [NebulousLabs/Sia-Nextcloud#6]
- upload keeps 0% until i restart siad
- have a name like `IMG_2777.jpg.ocTransferId1256960342.part` , can't see in nextcloud frontend
- after i rename it to `IMG_2777.jpg`,i can see it in the list,but can't download

**Comment by tlightsky** [comment]
try to download s.txt from sia in next cloud ,
this is printout of ` curl -A "Sia-Agent" http://localhost:9980/renter/downloads`
it seems try to download multi times and all files from sia,
and it didn't download successfully btw

```
{
	"downloads": [
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_p2jqVu-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:57:48.734932042+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_ik3iEA-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:56:25.62106846+09:00",
			"error": ""
		},
		{
			"siapath": "IMG_2777.jpg",
			"destination": "/tmp/oc_tmp_RttvJ5-.jpg",
			"filesize": 2105629,
			"received": 2105629,
			"starttime": "2017-04-17T12:56:25.454455056+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_MHgtNF-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:56:13.316264276+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_AZvmCm-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:56:02.74905349+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_zZuB8i-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:58.756238563+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_ue9gU1-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:53.720307076+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_q9iv8i-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:40.693329154+09:00",
			"error": ""
		},
		{
			"siapath": "IMG_2777.jpg",
			"destination": "/tmp/oc_tmp_eyN5uQ-.jpg",
			"filesize": 2105629,
			"received": 2105629,
			"starttime": "2017-04-17T12:55:30.348403135+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_OedWeg-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:25.954200585+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_uayWYq-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:16.675838339+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_30qE2n-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:15.639992666+09:00",
			"error": ""
		},
		{
			"siapath": "IMG_2777.jpg",
			"destination": "/tmp/oc_tmp_bxvtYg-.jpg",
			"filesize": 2105629,
			"received": 2105629,
			"starttime": "2017-04-17T12:55:13.149603749+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_PTPHk8-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:10.105248429+09:00",
			"error": ""
		},
		{
			"siapath": "IMG_2777.jpg",
			"destination": "/tmp/oc_tmp_XM1sPZ-.jpg",
			"filesize": 2105629,
			"received": 2105629,
			"starttime": "2017-04-17T12:55:07.108480758+09:00",
			"error": ""
		},
		{
			"siapath": "s.txt",
			"destination": "/tmp/oc_tmp_f7UmdT-.txt",
			"filesize": 205,
			"received": 205,
			"starttime": "2017-04-17T12:55:04.777654257+09:00",
			"error": ""
		}
	]
}
```

**Comment by tlightsky** [comment]
i try to download again, it would do 2 wrong thing i think:

- download one file for times,and would not download successfully 
- download other files which is not requested

@DavidVorick @johnathanhowell 

(upload seems work now, except some XXX.ext.ocXXXXX.part file in datadir)

**Comment by tlightsky** [comment]
read some code,
it seems that the nextcloud framework would do lib/Storage/Sia.fopen a alot of times which cause sia arrange many(at least 3) download work,
and it might be that download use a /tmp dir which casue the browser could not get the downloaded file

```
	public function fopen($path, $mode) {
		switch ($mode) {
			case 'r':
			case 'rb':
				$ext = pathinfo($path)['extension'];
				$tmpFile = \OCP\Files::tmpFile($ext);

				$this->client->download($path, $tmpFile);

				$hasDownloaded = false;
				while (!$hasDownloaded) {
					sleep(1);
					$downloads = $this->client->downloads();
					foreach($downloads as $download) {
						if ($download->destination == $tmpFile && $download->received == $download->filesize) {
							$hasDownloaded = true;
							break;
						}
					}
				}

				return fopen($tmpFile, 'r');
```

**Comment by avahowell** [comment]
Nextcloud does call fopen quite a lot, especially if you're uploading images (it will use `fopen` to get thumbnails for the images).

The solution to this is to check the local datadir first and return a handle to the file on disk if it exists, otherwise download it. I will push a fix for this soon.

**Comment by avahowell** [comment]
I was able to reproduce the problems here, they should be fixed as of 0.1.2.
